File: client/net/client/Com.Bj58.Spat.Gaea.Client/Configuration/Commmunication/ProtocolProfile.cs
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
//----------------------------------------------------------------------
//<Copyright company="58.com">
//      Team:SPAT
//      Blog:http://blog.58.com/spat/  
//      Website:http://www.58.com
//</Copyright>
//-----------------------------------------------------------------------
using System;
using System.Text;
using System.Xml;
using Com.Bj58.Spat.Gaea.Client.Communication.Protocol.SFP.Enum;
using Com.Bj58.Spat.Gaea.Client.Communication.Serialize;

namespace Com.Bj58.Spat.Gaea.Client.Configuration.Commmunication
{
    internal class ProtocolProfile
    {
        public ProtocolProfile(XmlNode node)
        { 
            XmlAttribute attrSer = node.Attributes["serialize"];
            if (attrSer == null)
                throw new XmlException("Not find attrbuts:" + node.Name + "[@'serialize']");
            string value = attrSer.Value.Trim().ToLower();
            if (value.Equals("json"))
            {
                this.SerializerType = SerializeType.JSON;
            }
            else if (value.Equals("binary"))
            {
                this.SerializerType = SerializeType.Binary;
            }
            else if (value.Equals("gaea"))
            {
                this.SerializerType = SerializeType.GAEA;
            }
            else
            {
                throw new NotSupportedException("Protocol not supported!");
            }
            XmlAttribute attrSerializer = node.Attributes["serializeType"];
            if (attrSerializer != null)
            {
                string serializeType = attrSerializer.Value;
                if (!string.IsNullOrEmpty(serializeType))
                {
                    string[] types = serializeType.Split(',');
                    if (types.Length == 2)
                    {
                       
                        this.Serializer = System.Reflection.Assembly.Load(types[0]).CreateInstance(types[1]) as SerializeBase;
                    }
                    else if (types.Length == 1)
                    {
                        this.Serializer = Activator.CreateInstance(Type.GetType(types[0])) as SerializeBase;
                    }
                }
            }
            if (Serializer == null)
            {
                this.Serializer = SerializeBase.GetInstance();
            }
            XmlAttribute attrEncoder = node.Attributes["encoder"];
            if (attrEncoder == null)
                this.Encoder = Encoding.UTF8;
            else
                this.Encoder = Encoding.GetEncoding(attrEncoder.Value);
            this.Serializer.Encoder = this.Encoder;
            ServiceID = byte.Parse(node.ParentNode.ParentNode.Attributes["id"].Value);
            Compress = (CompressType)Enum.Parse(typeof(CompressType), node.Attributes["compressType"].Value, true);
            
        }

        public SerializeType SerializerType
        {
            get;
            set;
        }
        public SerializeBase Serializer
        {
            get;
            set;
        }
        public Encoding Encoder
        {
            get;
            set;
        }
        public byte ServiceID
        {
            get;
            set;
        }
        public CompressType Compress
        {
            get;
            set;
        }
    }
}

File: client/java/client/src/com/bj58/spat/gaea/client/configuration/commmunication/ProtocolProfile.java
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package com.bj58.spat.gaea.client.configuration.commmunication;

import java.nio.charset.Charset;

import javax.naming.directory.NoSuchAttributeException;

import org.w3c.dom.Node;

import com.bj58.spat.gaea.protocol.serializer.SerializeBase;
import com.bj58.spat.gaea.protocol.sfp.enumeration.CompressType;
import com.bj58.spat.gaea.protocol.sfp.enumeration.SerializeType;

/**
 * ProtocolProfile
 *
 * @author Service Platform Architecture Team (spat@58.com)
 */
public class ProtocolProfile {

    private SerializeType serializerType;
    private SerializeBase serialize;
    public Charset Encoder;
    public byte serviceID;
    public CompressType compress;

    public ProtocolProfile(Node node) throws Exception {
        Node attrSer = node.getAttributes().getNamedItem("serialize");
        if (attrSer == null) {
            throw new ExceptionInInitializerError("Not find attrbuts:" + node.getNodeName() + "[@'serialize']");
        }
        String value = attrSer.getNodeValue().trim().toLowerCase();
        if (value.equalsIgnoreCase("binary")) {
            serializerType = SerializeType.JAVABinary;
        } else if (value.equalsIgnoreCase("json")) {
            serializerType = SerializeType.JSON;
        } else if (value.equalsIgnoreCase("xml")) {
            serializerType = SerializeType.XML;
        } else if (value.equalsIgnoreCase("gaea")) {
            serializerType = SerializeType.GAEABinary;
        } else {
            throw new NoSuchAttributeException("Protocol not supported " + value + "!");
        }
        this.serialize = SerializeBase.getInstance(serializerType);
        attrSer = node.getAttributes().getNamedItem("encoder");
        if (attrSer == null) {
            this.Encoder = Charset.forName("UTF-8");
        } else {
            this.Encoder = Charset.forName(attrSer.getNodeValue());
        }
        this.serialize.setEncoder(this.Encoder);
        serviceID = Byte.parseByte(node.getParentNode().getParentNode().getAttributes().getNamedItem("id").getNodeValue());//TODO 待检验
        compress = (CompressType) Enum.valueOf(CompressType.class, node.getAttributes().getNamedItem("compressType").getNodeValue());
    }

    public Charset getEncoder() {
        return Encoder;
    }

    public CompressType getCompress() {
        return compress;
    }

    public SerializeBase getSerializer() {
        return serialize;
    }

    public SerializeType getSerializerType() {
        return serializerType;
    }

    public byte getServiceID() {
        return serviceID;
    }
}
